tests/0113-cooperative_rebalance.cpp

```diff
@@ -26,227 +26,339 @@
+                     std::vector<RdKafka::TopicPartition*> &partitions) {
+    Test::Say(tostr() << "RebalanceCb: " << consumer->name() << " " << RdKafka::err2str(err) << ": " << part_list_print(partitions) << "\n");
     if (err == RdKafka::ERR__ASSIGN_PARTITIONS) {
-      consumer->incremental_assign(partitions);
-//      partition_cnt = (int)partitions.size();
+      RdKafka::Error *error = consumer->incremental_assign(partitions);
+      if (error)
+        Test::Fail(tostr() << "consumer->incremental_assign() failed: " << error->str());
+      assign_call_cnt += 1;
+      partitions_assigned_net += partitions.size();
     } else {
-      consumer->unassign();
-//      partition_cnt = 0;
+      if (consumer->assignment_lost())
+        lost_call_cnt += 1;
+      RdKafka::Error *error = consumer->incremental_unassign(partitions);
+      if (error)
+        Test::Fail(tostr() << "consumer->incremental_unassign() failed: " << error->str());
+      revoke_call_cnt += 1;
+      partitions_assigned_net -= partitions.size();
     }
-//    eof_cnt = 0;
   }
+
 };
 
 
-/** incremental assign, then assign(NULL)
+
+
+
+/* -------- a_assign_tests
+ *
+ * check behavior incremental assign / unassign outside the context of a rebalance.
+ */
+
+
+/** Incremental assign, then assign(NULL).
  */
-static void direct_assign_test_1(RdKafka::KafkaConsumer *consumer,
-                                 std::vector<RdKafka::TopicPartition *> toppars1,
-                                 std::vector<RdKafka::TopicPartition *> toppars2) {
+static void assign_test_1 (RdKafka::KafkaConsumer *consumer,
+                           std::vector<RdKafka::TopicPartition *> toppars1,
+                           std::vector<RdKafka::TopicPartition *> toppars2) {
   RdKafka::ErrorCode err;
   RdKafka::Error *error;
   std::vector<RdKafka::TopicPartition *> assignment;
 
-  test_assert(assignment.size() == 0, "Expecting current assignment to have size 0");
+  Test::Say("Incremental assign, then assign(NULL)\n");
+
+  if (assignment.size() != 0)
+    Test::Fail(tostr() << "Expecting current assignment to have size 0, not: " << assignment.size());
   if ((error = consumer->incremental_assign(toppars1))) {
-    Test::Fail("Incremental assign failed: " + error->str());
+    Test::Fail(tostr() << "Incremental assign failed: " << error->str());
     delete error;
   }
-  if ((err = consumer->assignment(assignment))) Test::Fail("Failed to get current assignment: " + RdKafka::err2str(err));
-  test_assert(assignment.size() == 1, "Expecting current assignment to have size 1");
+  if ((err = consumer->assignment(assignment)))
+    Test::Fail(tostr() << "Failed to get current assignment: " << RdKafka::err2str(err));
+  if (assignment.size() != 1)
+    Test::Fail(tostr() << "Expecting current assignment to have size 1, not: " << assignment.size());
   delete assignment[0];
   assignment.clear();
   if ((err = consumer->unassign())) Test::Fail("Unassign failed: " + RdKafka::err2str(err));
```

Comment:
Putting the Fail on the same line is muscle memory from my side project coding style. I find it easier to read because the key logic isn't muddled with the error handling visually, and each logical step is on a different line. but i'll put the fails on different lines, and put in some line gaps as well.